### ISSUE TITLE
Design discussion of the psa-crypto crate

### ISSUE THREAD

**Issue by hug-dev** [parallaxsecond/rust-psa-crypto#8]
I assume that the `psa-crypto-sys` crate would re-exports all of of the public items of PSA Crypto API, with their expected names (functions, types, macros, constants).

I propose the following design for the `psa-crypto` crate:
* in `src/lib.rs`: create the `PsaCrypto` struct and the `static ONCE: AtomicBool;`. A `new` method creates an instance, sets the `ONCE` bool to true, initialises the library or returns an error if already created
* `types/key.rs` containing `Attributes` (`psa_key_attributes_t`), `Lifetime`, `Id`, `Type`, `Usage` and `Policy` (an `Algorithm` and an `Usage`). Most of the types copied from [here](https://docs.rs/parsec-interface/0.14.0/parsec_interface/operations/psa_key_attributes/struct.KeyAttributes.html).
* `types/algorithm.rs` containing the [`Algorithm`](https://docs.rs/parsec-interface/0.14.0/parsec_interface/operations/psa_algorithm/enum.Algorithm.html) structure.
* `types/status.rs` containing the `Status` enumeration with all `psa_status_t`. Create a `Result` typedef with `Status` as the error type
* `operations/` containing the files defining all methods on `PsaCrypto`: `key_management.rs`, `message_digest.rs`, `mac.rs`, `ciphers.rs`, `aead.rs`, `key_derivation.rs`, `asym_signature.rs`, `asym_encryption.rs`, `key_agreement.rs` and `other_crypto.rs`.

Each one of the `types` files would implement conversions to and from the corresponding FFI type.

Create the function signatures applying the following systematic method:
* remove the `psa_` prefix and the `_t` suffix and use camel-case
* make all functions return a `Result`
* instead of the `psa_..._t` type, replace with the Rust native type. If they are small enough, it might be simpler to implement `Copy` on them and pass them directly instead of by reference
* when taking an input buffer (`const uint8_t *toto` and `size_t toto_length`) replace with a byte slice: `&[u8]`
* when returning a value (like a key ID), put it in the `Ok` variant of the `Result`
* when taking an output buffer (`uint8_t *toto`, `size_t toto_size` and `size_t *toto_length`) replace with a mutable byte slice `&mut [u8]` and return a `usize`. The reason for not using a `Vec<u8>` is that it would require `std` and would not be applicable on embedded targets. We could produce higher-level function with a `cfg` flag for when `std` is available, which use `Vec` for more simplicity.

## Safety

We have to provide a safe interface. Using only Rust types and assuming the safety of PSA Crypto when used correctly we can achieve memory safety easily.
Thread safety is more tricky and can add more challenges. Section 5.4 lists the constraints needed to achieve thread-safety:

> 1. There is no overlap between an output parameter of one call and an input or output parameter of another call. Overlap between input parameters is permitted.
> 2. If a call destroys a key, then no other call must destroy or use that key. Using, in this context, includes all functions of multi-part operations which have used the key as an input in a previous function.
> 3. Concurrent calls that use the same key are permitted.
> 4. Concurrent calls must not use the same operation object.

Rust safety guarantees, I think, prevent us from 1 and 4. 3 is not really a constraint.
2 is the trickiest and I am not sure how to handle it. If it is undefined behaviour to destroy a key more than once, we need to make sure that `destroy` is called only on `KeyId` of keys that exist. For persistent keys: how to retrieve, at instantiation, the keys that are already created?
I think the most simple way to approach this for now would be to put the `destroy` function as `unsafe`.

**Comment by hug-dev** [comment]
@ionut-arm @egrimley-arm Can I kindly ask for your review on this design?

**Comment by egrimley-arm** [comment]
What's the reason for having the `PsaCrypto` struct? What if there are two crates (A and B) that both use `psa-crypto`, and a third crate (C) that uses A and B. Then whichever of A and B grabs the `PsaCrypto` struct first succeeds, and the other one fails in an unrecoverable way?

**Comment by egrimley-arm** [comment]
Can 2 be solved with a readers-writer lock per key?

**Comment by egrimley-arm** [comment]
I don't see anything written about `psa_key_id_t`. Presumably a user of `psa-crypto` could foolishly create keys and discard the IDs until the underlying implementation runs out of slots. I don't think it's possible to prevent that from happening. To help a user understand what's happened, you could have the wrapper keep track of existing keys and add a debug function to list the existing keys. However, this would allow crate A (in the example above) to see crate B's keys, which would not be a security problem, but it would be a bit confusing.

**Comment by egrimley-arm** [comment]
It is also possible for the underlying implementation to leak resources by `psa_reset_key_attributes` not being called. Unlike with the keys, which are likely to be secret and can't be stored in the caller's memory, you have several options for implementing this. You could have every function that takes `Attributes` create a new `psa_key_attributes_t` and reset it afterwards, turning a single call to `psa-crypto` into a dozen or so calls to `psa-crypto-sys`. Or you could make the user of `psa-crypto` responsible for the resource management. Or you could cache these things in the wrapper. Or ... other options?

**Comment by hug-dev** [comment]
> 
> 
> What's the reason for having the `PsaCrypto` struct? What if there are two crates (A and B) that both use `psa-crypto`, and a third crate (C) that uses A and B. Then whichever of A and B grabs the `PsaCrypto` struct first succeeds, and the other one fails in an unrecoverable way?

I am taking the assumption that at some point, we will need to store information about the current state of the wrapper (for example to add thread safety). Maybe that assumption is false and if it is, I agree that this structure would be useless.
In your example, A and B would not need to call `new` but be passed a reference to `PsaCrypto` in some way. I agree that is not ideal.
Our other option is to put this data in global `lazy_static` structure, wrapped in a synchronisation structure like `Mutex` or `RwLock`.

> Can 2 be solved with a readers-writer lock per key?

It could! I think the important thing here is that we would have to keep the `KeyId` instances **unique**. If we do that, then we can use the borrow checker to ensure thread-safety: `generate_key` and `import_key` create a new instance of `KeyId`, all methods take a reference to it and `destroy_key` takes ownership. That way you can not destroy a key if any thread is currently using one of its references.
The problem with that are persistent keys: how to know, when starting the application, the persistent key IDs that already have been created?

> Presumably a user of psa-crypto could foolishly create keys and discard the IDs until the underlying implementation runs out of slots.

If we have the uniqueness of `KeyId` as described above, we could re-implement `Drop` so that the volatile keys gets destroyed when going out-of-scope. That would still happen for persistent keys but I think that is expected.

> It is also possible for the underlying implementation to leak resources by psa_reset_key_attributes not being called.

Reading at the API, `psa_key_attributes_t` seem to be short-lived, you create one, give it to `generate`/`import` and then reset it. Or, you get one from a KeyId, get the information that you need and then reset it.
This sounds like an excellent candidate for a `Drop` implementation on `Attributes` which calls `psa_reset_key_attributes` whenever it goes out of scope. That way, no risk to leak data.

**Comment by hug-dev** [comment]
> The problem with that are persistent keys: how to know, when starting the application, the persistent key IDs that already have been created?

I think we do not need to know at startup all the persistent keys that have been created but only when trying to create a `KeyId` from the ID number (`uint32_t`). At that moment, in a constructor of `KeyId`, like `from_persistent_key_id(id: u32)` we can check that a key of this ID actually exists with a call to `psa_generate_key` which is expected to fail with `PSA_ERROR_ALREADY_EXISTS`. If that the case, then we can create a valid `KeyId` structure.
To keep the uniqueness of the `KeyId` we create, we would store the persistent key IDs in a `HashSet` guarded with a `RwLock`.
`generate` and `import` would return a `KeyId` and add the ID to the set.
`destroy` would take ownership of the `KeyId` in parameter and remove it of the set.
All the other methods needing a key ID would take a simple reference to it.
In the `Drop` implementation of the `KeyId` structure we would check its lifetime, and if it is volatile, destroy the key but not if it is persistent.

**Comment by ionut-arm** [comment]
> It could! I think the important thing here is that we would have to keep the KeyId instances unique. If we do that, then we can use the borrow checker to ensure thread-safety: generate_key and import_key create a new instance of KeyId, all methods take a reference to it and destroy_key takes ownership. That way you can not destroy a key if any thread is currently using one of its references.
The problem with that are persistent keys: how to know, when starting the application, the persistent key IDs that already have been created?

You also run into a problem with multi-threading - how could two threads (or even the same thread, in different scopes) make use of the same key? Do they get an `Arc` of it? Or does the `KeyId` itself have an `Arc` in it with the handle inside? The problem would then be that you can't implement `Drop` on that `Arc`, but we can probably find some ways around that. Not sure if I'm just missing the point completely here, tho.

**Comment by ionut-arm** [comment]
> I think we do not need to know at startup all the persistent keys that have been created but only when trying to create a KeyId from the ID number (uint32_t). At that moment, in a constructor of KeyId, like from_persistent_key_id(id: u32) we can check that a key of this ID actually exists with a call to psa_generate_key which is expected to fail with PSA_ERROR_ALREADY_EXISTS. If that the case, then we can create a valid KeyId structure.

Is there no other way to check if such a key exists? I find this to be a weird omission from the API, tbh... Just as it does for Parsec - we don't allow people to create keys with the same name, but at the same time the keys are persistent and users can't discover them. 

Also, imagine if you're in the middle of that `from_persistent_key_id`, you've called `psa_generate_key` which returned success, and then your process gets killed - next time this runs you'll find a key there which was just created. Is this intended? 

> To keep the uniqueness of the KeyId we create, we would store the persistent key IDs in a HashSet guarded with a RwLock.
generate and import would return a KeyId and add the ID to the set.
destroy would take ownership of the KeyId in parameter and remove it of the set.
All the other methods needing a key ID would take a simple reference to it.
In the Drop implementation of the KeyId structure we would check its lifetime, and if it is volatile, destroy the key but not if it is persistent.

I'm not sure I follow why having both unique `KeyId` values **and** having a set with a `RwLock` on it is needed. I'm all for having the set that can be used as a central source of tracking which handles have been allocated.

**Comment by egrimley-arm** [comment]
Calling a function to generate a key while expecting it to fail, with that being an implicit part of the `KeyId` creation procedure, doesn't sound right to me. That isn't what most people would expect a `psa-crypto` "wrapper" to do. It's possible to keep a record of which keys are in use without doing that.

**Comment by hug-dev** [comment]
> You also run into a problem with multi-threading - how could two threads (or even the same thread, in different scopes) make use of the same key?

They would be given a reference to the `KeyId` which would be enough for them to execute all operations needing that key.

> Also, imagine if you're in the middle of that from_persistent_key_id, you've called psa_generate_key which returned success, and then your process gets killed - next time this runs you'll find a key there which was just created. Is this intended?

That's true and I am not sure  how to handle that. Worst case scenario it would be memory leak for something that should very rarely happen.

> I'm not sure I follow why having both unique KeyId values and having a set with a RwLock on it is needed. I'm all for having the set that can be used as a central source of tracking which handles have been allocated.

If you do not have the set you could call the `from_persistent_key_id` function indefinitely and that would break the uniqueness.

> Is there no other way to check if such a key exists? I find this to be a weird omission from the API, tbh... 

> Calling a function to generate a key while expecting it to fail, with that being an implicit part of the KeyId creation procedure, doesn't sound right to me. That isn't what most people would expect a psa-crypto "wrapper" to do. It's possible to keep a record of which keys are in use without doing that.

I agree this does not look good but I do not see any other way to check if a persistent Key ID, **not currently in use**, has already been created or not. If you see any other solutions please feel free to say.

The only other solutions I see are:
* storing in a file which persistent keys have been created and using that as a record. Maybe that is what you meant @egrimley-arm ?
* put the `delete` function as `unsafe`, state the problem clearly in the documentation and let the user create a safe wrapper around it with their own Key ID store.

**Comment by ionut-arm** [comment]
> They would be given a reference to the KeyId which would be enough for them to execute all operations needing that key.

So who is in fact the owner of that `KeyId`? Trying to figure out how this would work - would one thread have to send the references to another or would the crate/method have to store a reference and duplicate it?

> Worst case scenario it would be memory leak for something that should very rarely happen.

It's not really a memory leak since the key would be persisted (or at least that's how I understand it?). If this happens on an embedded system this opens the door for fault injection - reset the device at just the right time... True, it's unlikely to happen, but I'm just pointing it out. I don't have an alternative for how we could identify such keys. Maybe Gilles has some hints.

> If you do not have the set you could call the from_persistent_key_id function indefinitely and that would break the uniqueness.

I was advocating for having the set but not `KeyId` uniqueness.

**Comment by egrimley-arm** [comment]
I don't see why the wrapper has to check whether a persistent key ID has already been created. If the user attempts an operation on key 1234, say, an operation which might fail if the key exists when it shouldn't or doesn't exist when it should, then the wrapper just needs to record, only for the duration of that operation, that key 1234 is in use. If another thread tries to delete key 1234 during that operation then that thread will block until the other operation has finished.

**Comment by hug-dev** [comment]
> So who is in fact the owner of that KeyId? Trying to figure out how this would work - would one thread have to send the references to another or would the crate/method have to store a reference and duplicate it?

That one: "one thread have to send the references to another". If one thread creates a `KeyId` and wants that key to be used by multiple threads, he sends references to it to those threads.

> It's not really a memory leak since the key would be persisted (or at least that's how I understand it?).

Yes persistent memory leak 😋 

> I don't see why the wrapper has to check whether a persistent key ID has already been created.

This is to solve this constraint:

> If a call destroys a key, then no other call must destroy or use that key. Using, in this context, includes all functions of multi-part operations which have used the key as an input in a previous function.

and specifically this: "If a call destroys a key, then no other call must destroy that key." The `destroy_key` operation does not have a `PSA_ERROR_DOES_NOT_EXIST` return status so I supposed that calling `destroy_key` on an ID that was already destroyed or was never created was undefined behaviour. I don't think that having a record of keys in use would solve that problem because you could have two `destroy_key` calls on the same Key ID, one after the other triggering this issue.

> I was advocating for having the set but not KeyId uniqueness.

The uniqueness is I think an elegant way of using the Rust type system to maintain thread safety without having to record keys that are currently in use.

**Comment by egrimley-arm** [comment]
I would expect a call to `psa_destroy_key` to return `PSA_ERROR_INVALID_HANDLE` if the key has already been destroyed.

**Comment by ionut-arm** [comment]
> That one: "one thread have to send the references to another". If one thread creates a KeyId and wants that key to be used by multiple threads, he sends references to it to those threads.

But that implies that there will always be one thread that has some higher priority or precedence over all other threads - i.e. that thread **must** always terminate after the other ones to which it passes the references. So essentially this bakes in another constraint for users of this crate just because of this concept of uniqueness, that if you want two threads consuming Mbed Crypto services in parallel, you necessarily have a third one to own the keys and pass the references down. 

Even if those `KeyId`s are not unique, you can enforce the same kind of requirements (using the key with a immutable reference, deleting with passing ownership) by having a `HashMap` internally that maps key handle to current number of users. Whenever a user calls destroy, you decrement the counter - if the counter is now 0, actually destroy the key.

> I would expect a call to psa_destroy_key to return PSA_ERROR_INVALID_HANDLE if the key has already been destroyed.

Indeed, reading section 3.2.3 Key identifiers , it seems that should be the case. But those constraints also specified in the spec are a bit confusing.

> then the wrapper just needs to record, only for the duration of that operation, that key 1234 is in use.

I would've thought that multi-part operations might make that tricky but again, that section in the spec about key identifiers seems to hint at operations simply returning `INVALID_HANDLE` if the key has been destroyed. I guess the question is "can you create a key with the same handle, again?"

**Comment by hug-dev** [comment]
So we have been confirmed by the PSA Crypto team that a `psa_destroy_key` call on a key identifier that does not exist is defined behaviour and will return `PSA_ERROR_DOES_NOT_EXIST`.
The specific constraint means that there should be a readers-writer constraint on a key: at any given time any readers (for crypto operations) but one single writer (to destroy the key).

Knowing that, it is not longer needed to check if a persistent key ID has already been created in previous runs but we need to protect against the readers-writer problem.

Multiple solutions were given in this thread:
1. store a unique `RwLock` per key ID (but allow users to create as many Key ID instances they want), lock for reading when using a key, lock for writing when deleting a key
2. maintain uniqueness of key ID instances at creation (it would fail if a persistent `KeyId` with the same ID has already been created, using a `HashSet` to verify, or if the ID is the one of a volatile key). No locking would be needed for operations as thread safety would be checked statically with ownership rules.

1 has the problem that one thread could delete the key that another one is currently using (in a thread safe way). That would be solved (as @ionut-arm said) having a `HashMap` of counter per key and really deleting it when the count is 0. This seems a lot like re-implementing a reference counting like `Rc` or `Arc` which would be more natural to use (assuming uniqueness) when you would like to share the key amoung multiple threads.

2 has the disadventage that if two threads are trying to simultaneously create a `KeyId` from the same ID number, one of them will fail. A library that is using PSA Crypto in a multithreaded environment and wants to use a particular ID will either have to be given the `KeyID` at thread creation or create it in isolation.

**Comment by egrimley-arm** [comment]
Presumably in some implementations you will in any case potentially have to cope (avoid undefined behaviour) when a different process deletes a key you are using: https://armmbed.github.io/mbed-crypto/html/overview/goals.html#optional-isolation

**Comment by ionut-arm** [comment]
2 also has the disadvantage that a "worker thread" cannot destroy the key - the "master thread" that owns all the key handles has to join the worker threads, destroy the key and then start them up again (because I don't think you can have guarantees that there isn't a reference to the key somewhere in a worker). Also if the `KeyId` is unique and is somehow dropped without destroying the key, then the key is not usable anymore.

I think we need to decide what exactly the behaviour should be when `psa_destroy_key` is called. I get it, if the `KeyId` is unique then destroying a key is easy-peasy. But given all the caveats and restrictions that we raise by using such a strong uniqueness concept, it feels like we're sacrificing the overall experience of using the library in a multi-threaded environment for our benefit of not doing any work, just leveraging the compiler.

**Comment by hug-dev** [comment]
> Presumably in some implementations you will in any case potentially have to cope (avoid undefined behaviour) when a different process deletes a key you are using: https://armmbed.github.io/mbed-crypto/html/overview/goals.html#optional-isolation

Agreed and maybe trying to be more clever than we should be in our wrapper would prevent the crate from being used for some implementations.

> 2 also has the disadvantage that a "worker thread" cannot destroy the key - the "master thread" that owns all the key handles has to join the worker threads, destroy the key and then start them up again (because I don't think you can have guarantees that there isn't a reference to the key somewhere in a worker).

That is correct but I expect this to be only for persistent keys that needs to be shared amoung different threads. Worker threads can still create their own volatile keys and their own persistent keys (if they use a key ID that nobody is already using).

> destroy the key and then start them up again

Why that part though? I expect an use-case to be: a master thread create a persistent `KeyId` from a key that was provisioned in the PSA Crypto implementation or that he created before. It needs to use that key for computation that can be parallelised and give it to multiple worker threads. If it wants to delete the key, it waits for the threads to finish and then delete it.

> Also if the KeyId is unique and is somehow dropped without destroying the key, then the key is not usable anymore.

Why? You would be able to re-create the `KeyId` if it was previously dropped.

> I think we need to decide what exactly the behaviour should be when psa_destroy_key is called.

Yes I think so. A simpler solution for now would be to put the `destroy_key` method as `unsafe` and do not implement this thread safe logic at all.

**Comment by egrimley-arm** [comment]
> Yes I think so. A simpler solution for now would be to put the destroy_key method as unsafe and do not implement this thread safe logic at all.

Having an `unsafe` function in the API seems really rough. So a much better solution, that is almost as easy to implement, would be to serialise all operations on keys. Then everything is safe and the only disadvantage is inferior performance in some cases.

**Comment by hug-dev** [comment]
> Having an unsafe function in the API seems really rough.

I don't think so as long as the reasons for unsafety are clearly written in the documentation.

I agree that serialising all key operations is also an easy way to solve this problem, but that decision could also be taken at the caller level.

The more I think about it the solution 1. above seems like a good choice, even if a key that is currently used might get destroyed, that is not the responsability of this wrapper to handle that case. This wrapper should only aim to be the simplest safe messenger from Rust to PSA Crypto. At least for our current needs.

I do not think the implementation should be blocked any longer for this thread safety problem. I think I am going to get started on the parts we agreed on, probably put `destroy_key` as `unsafe` for now and we can think about it later when the first part is done.

**Comment by ionut-arm** [comment]
I'm ok with `destroy_key` as `unsafe` as long as it's clearly documented what the caller has to do to ensure safety.

**Comment by hug-dev** [comment]
Design has bene implementend 🛠️
